### ISSUE TITLE
Can't add routes in handler functions

### ISSUE THREAD

**Issue by jonathanunderwood** [aio-libs/aiohttp#1540]
## Long story short
I'd like to be able to add new routes inside handler functions. Unfortunately, at present this causes a `RuntimeError: Cannot register a resource into frozen router.` exception.

## Expected behaviour
New route is added

## Actual behaviour
500 error code is returned, and `RuntimeError: Cannot register a resource into frozen router.` exception is seen on the server side.

## Steps to reproduce
```python
import asyncio
from aiohttp import web

async def handle2(request):
    text = "Hello2222, yo"
    return web.Response(text=text)

async def handle(request):
    name = request.match_info.get('name', "Anonymous")
    text = "Hello, " + name
    request.app.router.add_get('/foo', handle2)
    return web.Response(text=text)

loop = asyncio.get_event_loop()

app = web.Application(loop=loop)

app.router.add_get('/', handle)
web.run_app(app)
```

Running this causes:

```python
$ python test.py
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
Error handling request
Traceback (most recent call last):
  File "/Users/pll740/Code/aiohttp-test/lib/python3.5/site-packages/aiohttp/web_server.py", line 61, in handle_request
    resp = yield from self._handler(request)
  File "/Users/pll740/Code/aiohttp-test/lib/python3.5/site-packages/aiohttp/web.py", line 249, in _handle
    resp = yield from handler(request)
  File "/Users/pll740/Code/aiohttp-test/lib/python3.5/site-packages/aiohttp/web_urldispatcher.py", line 111, in handler_wrapper
    result = old_handler(*args, **kwargs)
  File "test.py", line 11, in handle
    request.app.router.add_get('/foo', handle2)
  File "/Users/pll740/Code/aiohttp-test/lib/python3.5/site-packages/aiohttp/web_urldispatcher.py", line 868, in add_get
    return self.add_route(hdrs.METH_GET, *args, **kwargs)
  File "/Users/pll740/Code/aiohttp-test/lib/python3.5/site-packages/aiohttp/web_urldispatcher.py", line 831, in add_route
    resource = self.add_resource(path, name=name)
  File "/Users/pll740/Code/aiohttp-test/lib/python3.5/site-packages/aiohttp/web_urldispatcher.py", line 795, in add_resource
    self._reg_resource(resource)
  File "/Users/pll740/Code/aiohttp-test/lib/python3.5/site-packages/aiohttp/web_urldispatcher.py", line 770, in _reg_resource
    raise RuntimeError("Cannot register a resource into "
RuntimeError: Cannot register a resource into frozen router.
```

## Your environment
aiobotocore==0.0.6
aiohttp==1.2.0
async-timeout==1.1.0
botocore==1.4.73
chardet==2.3.0
docutils==0.13.1
jmespath==0.9.0
multidict==2.1.4
python-dateutil==2.6.0
six==1.10.0
yarl==0.8.1

arwin XXXXXXXXX 15.6.0 Darwin Kernel Version 15.6.0: Wed Nov  2 20:30:56 PDT 2016; root:xnu-3248.60.11.1.2~2/RELEASE_X86_64 x86_64

**Comment by jonathanunderwood** [comment]
I see that doing this makes it possible to add a route:

```python
    request.app.router._frozen = False
    request.app.router.add_get('/foo', handle2)
    request.app.router._frozen = True
```

However, that feels like a hack. What's the reasoning behind freezing the routing/UrlDispatcher instance when the application is running?

**Comment by fafhrd91** [comment]
what is the use case for this?

**Comment by jonathanunderwood** [comment]
> what is the use case for this?

I want to have an endpoint, say /register, where POSTing to that triggers the creation of a new route (eg. /topics/topic1) for POSTing specific topic data to.

**Comment by fafhrd91** [comment]
why can't you just register the route upfront?

**Comment by jonathanunderwood** [comment]
Because I don't know what routes will be required when I start the app. I want to be able to create topic routes while the app is running in response to requests.

**Comment by fafhrd91** [comment]
you can register route that can handle all possible paths and then do custom routing based on custom rules. another option is custom dispatcher.

in any case registering route at runtime is bad idea, dispatcher freezing is explicit decision and we won't revert it.

**Comment by jonathanunderwood** [comment]
> dispatcher freezing is explicit decision

Is there any documented rationale for that?

**Comment by fafhrd91** [comment]
i don't think so

**Comment by jonathanunderwood** [comment]
OK, thanks. I'll use Tornado then.

**Comment by yope** [comment]
Sorry to comment on a close bug, but I just came across this while porting my application from aiohttp 0.22 to 1.2.
I have the same situation as OP: When starting the application, I used to add web-socket routes dynamically and cannot do that anymore.
If there is a reason for registering a route at run-time being a bad idea, I would also really like to know why? A mention of this in the documentation would of course also be fine.
